File: latency/latency.cpp

#include <fstream>
#include <iostream>
#include <string>
#include <vector>
#include <unistd.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>

#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <simplefso.h>

int makeUDPSocket(int port){
	// Make socket
	int sock = socket(AF_INET,SOCK_DGRAM,0);
	if(sock < 0){
		std::cerr << "Couldn't open socket" << std::endl;
		exit(0);
	}

	// Bind socket
	sockaddr_in addr;
	memset((void *)&addr, 0, sizeof(addr));
	addr.sin_family = AF_INET;
	addr.sin_addr.s_addr = htonl(INADDR_ANY);
	addr.sin_port = htons(port);

	int rv = bind(sock, (struct sockaddr *)&addr, sizeof(addr));
	if(rv < 0){
		std::cerr << "Couldn't bind socket" << std::endl;
		exit(0);
	}

	return sock;
}

void processCommandLine(int argc, char const *argv[],int &tp,int &nt,int &ps,float &wt,std::string &of,std::string &cf) {
	for(int i = 1; i < argc; ++i) {
		std::string val = argv[i];
		if((val.compare("-tp") == 0 || val.compare("--test_port") == 0) && i < argc - 1){
			i++;
			tp = atoi(argv[i]);
		}
		if((val.compare("-w") == 0 || val.compare("--wait_time") == 0) && i < argc - 1){
			i++;
			wt = atof(argv[i]);
		}
		if((val.compare("-nt") == 0 || val.compare("--num_test") == 0) && i < argc - 1){
			i++;
			nt = atoi(argv[i]);
		}
		if((val.compare("-ps") == 0 || val.compare("--packet_size") == 0) && i < argc - 1){
			i++;
			ps = atoi(argv[i]);
		}
		if((val.compare("-out") == 0 || val.compare("--out_file") == 0) && i < argc - 1){
			i++;
			of = argv[i];
		}
		if((val.compare("-config") == 0 || val.compare("--config_file") == 0) && i < argc - 1){
			i++;
			cf = argv[i];
		}
	}
}

int main(int argc, char const *argv[]) {
	int test_port = 9999;
	int num_test = 1;
	int packet_size = 1;
	float wait_time = 1.0; // in seconds
	std::string out_file = "";
	std::string config_file = "";

	processCommandLine(argc,argv,test_port,num_test,packet_size,wait_time,out_file,config_file);

	int micro_wait_time = int(wait_time * 1000000);

	// Make a socket
	int sock = makeUDPSocket(test_port);

	int buffer_size = packet_size + 10;
	char buffer[buffer_size];
	sockaddr_in from_addr;
	socklen_t addrlen = sizeof(from_addr);

	// read in config file
	SimpleFSO *fso = new SimpleFSO(config_file);

	fso->connectGM();

	std::vector<double> latency_vals;
	for(int i = 0; i < num_test; ++i){
		// Wait some time, this is to ensure that the gm doesn't switch too fast and break
		usleep(micro_wait_time);

		// Wait till recieve a packet
		recvfrom(sock,buffer,buffer_size,0,(sockaddr *)&from_addr,&addrlen);

		// record time
		const clock_t begin_time = clock();

		// Switch
		fso->switchLink();

		// recieve new packet
		recvfrom(sock,buffer,buffer_size,0,(sockaddr *)&from_addr,&addrlen);
		// record time
		const clock_t end_time = clock();  // Not most accurate,  at most accurate to .01
		latency_vals.push_back( double(end_time - begin_time) / CLOCKS_PER_SEC );
	}

	// Write data to file
	std::ofstream out;
	out.open(out_file.c_str(),std::ios::app);

	// Print params
	out << "packet_size=" << packet_size << ",wait_time=" << wait_time << ",language=C++";
	// Print data
	for(unsigned int i = 0; i < latency_vals.size(); ++i){
		out << " " << latency_vals[i];
	}
	out << std::endl;
	
	return 0;
}

File: latency/latency.py

from gm import GM
import argparse
import time
from socket import *
from datetime import datetime

class SimpleFSO:
	def __init__(self,config_file):
		self.val1 = None
		self.val2 = None
		self.is_val1 = True
		self.readConfig(config_file)
		self.gm = GM()
		self.gm.connectDevice()
		self.gm.setValue(self.val1)

	def readConfig(self,config_file):
		gm_fso_id = None
		fso_1_id = None
		fso_2_id = None
		vals = dict()
		f = open(config_file)
		for line in f:
			vs = line.split()
			if vs[0] == 'FSO':
				if len(vs) == 3 and vs[2] == 'GM':
					if gm_fso_id == None:
						gm_fso_id = vs[1]
					else:
						raise Exception('Incorrect network configuration')
				else:
					if fso_1_id == None:
						fso_1_id = vs[1]
					elif fso_2_id == None:
						fso_2_id = vs[1]
					else:
						raise Exception('Incorrect network configuration')
			if vs[0] == 'Link':
				vals[vs[1]] = int(vs[2])
		if None in [gm_fso_id,fso_1_id,fso_2_id]:
			raise Exception('Incorrect network configuration')
		self.val1 = vals[fso_1_id]
		self.val2 = vals[fso_2_id]

	def switchLink(self):
		if self.is_val1:
			self.gm.setValue(self.val2)
		else:
			self.gm.setValue(self.val1)
		self.is_val1 = not self.is_val1

def runTests(port,c1,c2,wait_time,num_test,packet_size,angle,out_file,config_file):
	sock = socket(AF_INET,SOCK_DGRAM)
	sock.bind(('',port))

	fso = SimpleFSO(config_file)

	vals = []
	for x in range(num_test):
		time.sleep(wait_time)

		data,addr = sock.recvfrom(packet_size + 100)
		if data[0] == c1:
			first_c1 = True
		else:
			first_c1 = False

		start_time = datetime.now()

		fso.switchLink()
		while True:
			data,addr = sock.recvfrom(packet_size + 100)
			if (data[0] == c1 and (not first_c1)) or (data[0] == c2 and first_c1):
				break

		end_time = datetime.now()

		vals.append((end_time - start_time).total_seconds())
	f = open(out_file,'a')

	f.write('packet_size=' + str(packet_size) + ',wait_time=' + str(wait_time) + ',language=python,angle=' + str(angle) + ' ')
	f.write(' '.join(map(str,vals)))
	f.write('\n')

if __name__ == '__main__':
	parser = argparse.ArgumentParser(description='Runs the receiver end of the througput test')

	parser.add_argument('-tp','--test_port',metavar = 'TP',type = int,nargs = 1,default = [9999],help = 'Local port used for test')
	parser.add_argument('-w','--wait_time',metavar = 'WT',type = float,nargs = 1,default = [1.0],help = 'time between tests')
	parser.add_argument('-nt','--num_test',metavar = 'NT',type = int,nargs = 1,default = [1],help = 'Number of tests to run')
	parser.add_argument('-ps','--packet_size',metavar = 'PS',type = int,nargs = 1,default = [1],help = 'Size of packet to be recieved')
	parser.add_argument('-out','--out_file',metavar = 'OF',type = str,nargs = 1,default = [''],help = 'File to output data')
	parser.add_argument('-config','--config_file',metavar = 'CF',type = str,nargs = 1,default = [''],help = 'File that has value which configure network')
	
	parser.add_argument('-c1','--char1',metavar = 'C1',type = str,nargs = 1,default = ['A'],help = 'Character from endpoint 1')
	parser.add_argument('-c2','--char2',metavar = 'C2',type = str,nargs = 1,default = ['A'],help = 'Character from endpoint 2')

	parser.add_argument('-a','--gm_angle',metavar = 'TP',type = int,nargs = 1,default = [-1],help = 'Angle GM rotates')
	

	args = parser.parse_args()

	angle = args.gm_angle[0]

	port = args.test_port[0]
	wait_time = args.wait_time[0]
	num_test = args.num_test[0]
	packet_size = args.packet_size[0]
	out_file = args.out_file[0]
	config_file = args.config_file[0]

	c1 = args.char1[0]
	c2 = args.char2[0]

	runTests(port,c1,c2,wait_time,num_test,packet_size,angle,out_file,config_file)
